env/kube_registry.go

```diff
@@ -0,0 +1,331 @@
+	if err != nil {
+		return nil, err
+	}
+
+	version := unversioned.GroupVersion{
+		Group:   IstioAPIGroup,
+		Version: IstioResourceVersion,
+	}
+
+	config.GroupVersion = &version
+	config.APIPath = "/apis"
+	config.ContentType = runtime.ContentTypeJSON
+	config.NegotiatedSerializer = serializer.DirectCodecFactory{CodecFactory: api.Codecs}
+
+	schemeBuilder := runtime.NewSchemeBuilder(
+		func(scheme *runtime.Scheme) error {
+			scheme.AddKnownTypes(
+				version,
+				&api.ListOptions{},
+				&api.DeleteOptions{},
+			)
+			for kind := range km {
+				scheme.AddKnownTypeWithName(unversioned.GroupVersionKind{
+					Group:   IstioAPIGroup,
+					Version: IstioResourceVersion,
+					Kind:    kind,
+				}, &Config{})
+				scheme.AddKnownTypeWithName(unversioned.GroupVersionKind{
+					Group:   IstioAPIGroup,
+					Version: IstioResourceVersion,
+					Kind:    kind + "List",
+				}, &ConfigList{})
+			}
+			return nil
+		})
+	schemeBuilder.AddToScheme(api.Scheme)
+
+	return config, nil
+}
+
+func NewKubernetesRegistry(kubeconfig string, km model.KindMap) (*KubernetesRegistry, error) {
+	config, err := CreateRESTConfig(kubeconfig, km)
+	if err != nil {
+		return nil, err
+	}
+	cl, err := kubernetes.NewForConfig(config)
+	if err != nil {
+		return nil, err
+	}
+
+	dyn, err := rest.RESTClientFor(config)
+	if err != nil {
+		return nil, err
+	}
+
+	out := &KubernetesRegistry{
+		mapping:   km,
+		client:    cl,
+		dyn:       dyn,
+		Namespace: api.NamespaceDefault,
+	}
+
+	if err = out.RegisterResources(); err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// RegisterResources creates third party resources
+func (kr *KubernetesRegistry) RegisterResources() error {
+	var out error
+	for kind, v := range kr.mapping {
+		apiName := KindToAPIName(kind)
+		// initialize resource if it does not exist
+		res, err := kr.client.Extensions().ThirdPartyResources().Get(apiName)
+		if err != nil {
+			if errors.IsNotFound(err) {
+				tpr := &v1beta1.ThirdPartyResource{
+					ObjectMeta: v1.ObjectMeta{
+						Name: apiName,
+					},
+					Versions: []v1beta1.APIVersion{
+						{Name: IstioResourceVersion},
+					},
+					Description: v.Description,
+				}
+				_, err = kr.client.Extensions().ThirdPartyResources().Create(tpr)
+				if err != nil {
+					out = multierror.Append(out, err)
+				}
+			} else {
+				out = multierror.Append(out, err)
+			}
+		} else {
+			log.Printf("Resource already exists: %s", res.Name)
+		}
+	}
+	return out
+}
+
+// DeregisterResources removes third party resources
+func (kr *KubernetesRegistry) DeregisterResources() error {
+	var out error
+	for kind := range kr.mapping {
+		apiName := KindToAPIName(kind)
+		err := kr.client.Extensions().ThirdPartyResources().Delete(apiName, &api.DeleteOptions{})
+		if err != nil {
+			out = multierror.Append(out, err)
+		}
+	}
+	return out
+}
+
+// CamelCaseToKabobCase converts "MyName" to "my-name"
+func CamelCaseToKabobCase(s string) string {
```

Comment:
i feel like this could be written using strings.Map, a la [strings.Title](https://golang.org/src/strings/strings.go?s=13125:13152#L475).